amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala

```diff
@@ -0,0 +1,259 @@
+                        hostAndPorts: (String, Int)*): AmqpConnectionFactoryConnectionProvider =
+    copy(hostAndPortList = (hostAndPort +: hostAndPorts).toList)
+
+  /**
+   * Java API
+   */
+  @annotation.varargs
+  def withHostsAndPorts(hostAndPort: akka.japi.Pair[String, Int],
+                        hostAndPorts: akka.japi.Pair[String, Int]*): AmqpConnectionFactoryConnectionProvider =
+    copy(hostAndPortList = (hostAndPort +: hostAndPorts).map(_.toScala).toList)
+
+  override def get: Connection = {
+    import scala.collection.JavaConverters._
+    factory.newConnection(hostAndPortList.map(hp => new Address(hp._1, hp._2)).asJava)
+  }
+}
+
+object AmqpConnectionFactoryConnectionProvider {
+
+  /**
+   * Java API
+   */
+  def create(factory: ConnectionFactory): AmqpConnectionFactoryConnectionProvider =
+    AmqpConnectionFactoryConnectionProvider(factory: ConnectionFactory)
+}
+
+final case class AmqpCachedConnectionProvider(provider: AmqpConnectionProvider, automaticRelease: Boolean = true)
+    extends AmqpConnectionProvider {
+
+  import akka.stream.alpakka.amqp.AmqpCachedConnectionProvider._
+  private val state = new AtomicReference[State](Empty)
+
+  @tailrec
+  override def get: Connection = state.get match {
+    case Empty =>
+      if (state.compareAndSet(Empty, Connecting)) {
+        val connection = provider.get
+        if (!state.compareAndSet(Connecting, Connected(connection, 1)))
+          throw new ConcurrentModificationException(
+            "Unexpected concurrent modification while creating the connection."
+          )
+        connection
+      } else get
+    case Connecting => get
+    case c @ Connected(connection, clients) =>
+      if (state.compareAndSet(c, Connected(connection, clients + 1))) connection
+      else get
+    case Closing => get
+    case unknownState => throw new IllegalStateException(s"Unknown provider state ($unknownState).")
+  }
+
+  @tailrec
+  override def release(connection: Connection): Unit = state.get match {
+    case Empty => throw new IllegalStateException("There is no connection to release.")
+    case Connecting => release(connection)
+    case c @ Connected(cachedConnection, clients) =>
+      if (cachedConnection != connection)
+        throw new IllegalArgumentException("Can't release a connection that's not owned by this provider")
+
+      if (clients == 1 || !automaticRelease) {
+        if (state.compareAndSet(c, Closing)) {
+          provider.release(connection)
+          if (!state.compareAndSet(Closing, Empty))
+            throw new ConcurrentModificationException(
+              "Unexpected concurrent modification while closing the connection."
+            )
+        }
+      } else {
+        if (!state.compareAndSet(c, Connected(cachedConnection, clients - 1))) release(connection)
+      }
+    case Closing => release(connection)
+    case unknownState => throw new IllegalStateException(s"Unknown provider state ($unknownState).")
```

Comment:
Ah, now I see why you wrote "unknown" here, as state is sealed we don't actually have any unknown states, we only need to cover all the possible ones in the ADT, so you can remove the `catch whatever =>` part from both open and release